judge/contest_format/ioi.py

```diff
@@ -79,15 +79,19 @@ def update_participation(self, participation):
     def display_user_problem(self, participation, contest_problem):
         format_data = (participation.format_data or {}).get(str(contest_problem.id))
         if format_data:
-            return format_html(
-                '<td class="{state}"><a href="{url}">{points}<div class="solving-time">{time}</div></a></td>',
-                state=(('pretest-' if self.contest.run_pretests_only and contest_problem.is_pretested else '') +
-                       self.best_solution_state(format_data['points'], contest_problem.points)),
-                url=reverse('contest_user_submissions',
-                            args=[self.contest.key, participation.user.user.username, contest_problem.problem.code]),
-                points=floatformat(format_data['points']),
-                time=nice_repr(timedelta(seconds=format_data['time']), 'noday') if self.config['cumtime'] else '',
-            )
+            try:
+                return format_html(
+                    '<td class="{state}"><a href="{url}">{points}<div class="solving-time">{time}</div></a></td>',
+                    state=(('pretest-' if self.contest.run_pretests_only and contest_problem.is_pretested else '') +
+                           self.best_solution_state(format_data['points'], contest_problem.points)),
+                    url=reverse('contest_user_submissions',
+                                args=[self.contest.key, participation.user.user.username,
+                                      contest_problem.problem.code]),
+                    points=floatformat(format_data['points']),
+                    time=nice_repr(timedelta(seconds=format_data['time']), 'noday') if self.config['cumtime'] else '',
+                )
+            except (KeyError, TypeError, ValueError):
+                return mark_safe('<td>???</td>')
```

Comment:
Rather than copying this over and over, perhaps it should be up to the caller of `display_user_problem`?